apis/pipelines/hub/pipeline_types.go

```diff
@@ -26,25 +26,76 @@ type PipelineFramework struct {
 	Parameters map[string]*apiextensionsv1.JSON `json:"parameters" yaml:"parameters"`
 }
 
+const FallbackFramework = "tfx"
+
 func NewPipelineFramework(compilerType string) PipelineFramework {
 	return PipelineFramework{
 		Name:       compilerType,
 		Parameters: make(map[string]*apiextensionsv1.JSON),
 	}
 }
 
+/*
+ComputeHash returns a hash of the pipeline spec, ensuring consistency across versions.
+
+To maintain compatibility, certain fields are deliberately excluded from the hash calculation, or are normalized to match
+the behaviour of previous versions when using default fields as determined by conversion to the hub version.
+
+The order in which data is written to the hash is critical and must remain unchanged across versions to ensure consistency.
+
+This consistency across versions is preferable as it will prevent runs being triggered for pipelines that have only changed
+semantically without any spec changes, such as during an api version upgrade.
+*/
 func (ps Pipeline) ComputeHash() []byte {
+	pipeline := ps.DeepCopy()
 	oh := pipelines.NewObjectHasher()
-	oh.WriteStringField(ps.Spec.Framework.Name)
+	oh.WriteStringField(ps.Spec.Image)
+
+	isFallbackFramework := pipeline.Spec.Framework.Name == FallbackFramework
 
-	output := make(map[string]string)
-	for key, value := range ps.Spec.Framework.Parameters {
-		raw, _ := json.Marshal(value.Raw)
-		output[key] = string(raw)
+	// For the default framework, include components in the hash using the same format as spoke versions to ensure consistency.
+	tfxComponentsString := ""
+	if isFallbackFramework && pipeline.Spec.Framework.Parameters["components"] != nil {
+		if err := json.Unmarshal(pipeline.Spec.Framework.Parameters["components"].Raw, &tfxComponentsString); err != nil {
+			return nil
+		}
```

Comment:
nit: Rather than discard the error from unmarshalling the components meta, maybe add a debug log (note similar instances below too).

Reply:
I had considered this, exposing a log to identify there is an issue would be beneficial. There are drawbacks though to consider:

There currently is no logger exposed within this struct, and the interface it is adhering to for this computeVersion() function does not permit a logger or context to be passed through as a parameter.

Updating this will mean updating  the interface, in turn all implementations and many of the call sites will need reworking to expose the context also. It should be enough in this case to rely on the hash not being calculated to identify an issue for now 😃